### PR TITLE
CreateApiKey REST example with OAuth.

### PR DESCRIPTION
**Pull request by aamirbutt**

### CODE REVIEW

**Comment by tallatshafaat** [comment]
@lgoodby kindly be the main reviewer. Thanks.

File: feature-examples/python/create_apikey.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import vectara_util
+
+
+def create_apikey(
+    customer_id: int,
+    corpus_id: int,
+    auth_url: str,
+    app_client_id: str,
+    app_client_secret: str,
+):
```
Comment (lgoodby) [inline feature-examples/python/create_apikey.py:18]: if using type annotations, suggest specifying the return type as well.

File: feature-examples/python/create_apikey.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                "description": "API Key for testing.",
+                "apiKeyType": 2,  # 1 - Query, 2 - Query & Indexing
+                "corpusId": [corpus_id],  # One key can be used for multiple corpora.
+            }
+        ]
+    }
+
+    response = requests.post(
+        f"https://api.vectara.io/v1/create-api-key",
+        data=json.dumps(request),
```
Comment (lgoodby) [inline feature-examples/python/create_apikey.py:55]: you could do `json=request` instead.

File: feature-examples/python/create_apikey.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        help="Corpus ID to which API key will be created.",
+    )
+    parser.add_argument(
+        "--app-client-id",
+        required=True,
+        help="This app client should have enough rights.",
+    )
+    parser.add_argument("--app-client-secret", required=True)
+    parser.add_argument(
+        "--auth-url", required=True, help="The cognito auth url for this customer."
```
Comment (lgoodby) [inline feature-examples/python/create_apikey.py:107]: Please remove "cognito", as that is an internal implementation detail.

File: feature-examples/python/create_apikey.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    if message["response"]:
+        if len(message["response"]) != 1:
+            logging.error("CreateApiKey failed with response %s", message["response"])
+            return message["response"], False
+        status = message["response"][0]["status"]
+        if status["code"] != "OK":
+            logging.error("CreateApiKey failed with status %s", status)
+            return status, False
+
+        return message["response"][0]["keyId"], True
```
Comment (lgoodby) [inline feature-examples/python/create_apikey.py:79]: insert blank line after large outer `if` block.

**Review by lgoodby — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+.vscode/
```
[the remark this answers is not in the document]
Comment (aamirbutt) [reply .gitignore]: Done.

File: feature-examples/python/create_apikey.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    Args:
+        customer_id: Unique customer ID in vectara platform.
+        corpus_id: Corpus ID to which API key will be created.
+        auth_url: The cognito auth url for this customer.
+        app_client_id: The id of the app client with enough rights.
+        app_client_secret: The app client secret.
+
+    Returns:
+        (response, True) in case of success and returns (error, False) in case of failure.
+
```
[the remark this answers is not in the document]
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Done.

File: feature-examples/python/create_apikey.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        status = message["response"][0]["status"]
+        if status["code"] != "OK":
+            logging.error("CreateApiKey failed with status %s", status)
+            return status, False
+
+        return message["response"][0]["keyId"], True
+    return message, False
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Done.

File: feature-examples/python/create_apikey.py (hunk 4 above)
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Done.

File: feature-examples/python/create_apikey.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        )
+        return response, False
+
+    message = response.json()
+    if message["response"]:
+        if len(message["response"]) != 1:
+            logging.error("CreateApiKey failed with response %s", message["response"])
+            return message["response"], False
+        status = message["response"][0]["status"]
+        if status["code"] != "OK":
```
[the remark this answers is not in the document]
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Done.

File: feature-examples/python/create_apikey.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
+"""Example of using the Vectara REST API to create an API Key."""
+
+import argparse
+import logging
+import requests
+import sys
```
Comment (lgoodby) [inline feature-examples/python/create_apikey.py:6]: `requests` is third-party, so please move down to a new import block (standard imports only in first block, followed by third-party imports, followed by local imports, see https://peps.python.org/pep-0008/#imports).

File: feature-examples/python/create_apikey.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+    Args:
+        customer_id: Unique customer ID in vectara platform.
+        corpus_id: Corpus ID to which API key will be created.
+        auth_url: The cognito auth url for this customer.
+        app_client_id: The id of the app client with enough rights.
+        app_client_secret: The app client secret.
+
+    Returns:
+        (response, True) in case of success and returns (error, False) in case of failure.
```
Comment (lgoodby) [inline feature-examples/python/create_apikey.py:28]: Please clarify what `response` is. `(apikey, True)`?

**Review by lgoodby — CHANGES_REQUESTED** [review]

File: feature-examples/python/create_apikey.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            {
+                "description": "API Key for testing.",
+                "apiKeyType": 2,  # 1 - Query, 2 - Query & Indexing
+                "corpusId": [corpus_id],  # One key can be used for multiple corpora.
+            }
+        ]
+    }
+
+    response = requests.post(
+        f"https://api.vectara.io/v1/create-api-key",
```
[the remark this answers is not in the document]
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Removed f-string.

File: feature-examples/python/create_apikey.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    message = response.json()
+    if message["response"]:
+        if len(message["response"]) != 1:
+            logging.error("CreateApiKey failed with response %s", message["response"])
+            return message["response"], False
+        status = message["response"][0]["status"]
+
+        if status["code"] == "OK":
+            return message["response"][0]["keyId"], True
+        else:
```
[the remark this answers is not in the document]
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Done.

File: feature-examples/python/create_apikey.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+import vectara_util
+
+
+def create_apikey(
+    customer_id: int,
+    corpus_id: int,
+    auth_url: str,
+    app_client_id: str,
+    app_client_secret: str,
+) -> (any, bool):
```
[the remark this answers is not in the document]
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Done.

File: feature-examples/python/create_apikey.py (hunk 10 above)
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Done.

File: feature-examples/python/create_apikey.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        help="This app client should have enough rights.",
+    )
+    parser.add_argument("--app-client-secret", required=True)
+    parser.add_argument(
+        "--auth-url", required=True, help="The auth url for this customer."
+    )
+
+    args = parser.parse_args()
+
+    if args:
```
Comment (lgoodby) [inline feature-examples/python/create_apikey.py:115]: No need for `if args` here.

**Review by lgoodby — CHANGES_REQUESTED** [review]

File: feature-examples/python/create_apikey.py [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+        if status["code"] == "OK":
+            return message["response"][0]["keyId"], True
+        logging.error("CreateApiKey failed with status %s", status)
+        return status, False
+
+    return message, False
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Done.

File: feature-examples/python/create_apikey.py (hunk 14 above)
Comment (aamirbutt) [reply feature-examples/python/create_apikey.py]: Removed.

File: feature-examples/python/create_apikey.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+            response.text,
+        )
+        return response, False
+
+    message = response.json()
+    if message["response"]:
+        if len(message["response"]) != 1:
+            logging.error("CreateApiKey failed with response %s", message["response"])
+            return message["response"], False
+        status = message["response"][0]["status"]
```
Comment (lgoodby) [inline feature-examples/python/create_apikey.py:75]: nit: swap lines 75 and 76 so blank line comes after the `len != 1` block and `status` assignment is grouped with the `if status` block.

**Review by lgoodby — APPROVED** [review]
LGTM, thank you!

**Pull request merged**